modules/apps/oauth2-provider/oauth2-provider-rest/src/main/java/com/liferay/oauth2/provider/rest/internal/endpoint/liferay/LiferayOAuthDataProvider.java

```diff
@@ -1028,40 +1032,46 @@ private void _customizeServerAccessToken(
 	private Collection<LiferayOAuth2Scope> _getLiferayOAuth2Scopes(
 		long oAuth2ApplicationScopeAliasesId, List<String> scopeAliases) {
 
-		Collection<OAuth2ScopeGrant> oAuth2ScopeGrants =
-			_oAuth2ScopeGrantLocalService.getOAuth2ScopeGrants(
-				oAuth2ApplicationScopeAliasesId, QueryUtil.ALL_POS,
-				QueryUtil.ALL_POS, null);
-
-		Stream<OAuth2ScopeGrant> stream = oAuth2ScopeGrants.stream();
+		Collection<LiferayOAuth2Scope> liferayOAuth2Scopes = new ArrayList<>();
 
 		OAuth2ApplicationScopeAliases oAuth2ApplicationScopeAliases =
 			_oAuth2ApplicationScopeAliasesLocalService.
 				fetchOAuth2ApplicationScopeAliases(
 					oAuth2ApplicationScopeAliasesId);
 
-		Collection<LiferayOAuth2Scope> liferayOAuth2Scopes = new ArrayList<>();
-
 		if (oAuth2ApplicationScopeAliases != null) {
 			liferayOAuth2Scopes = _scopeLocator.getLiferayOAuth2Scopes(
 				oAuth2ApplicationScopeAliases.getCompanyId());
 		}
 
-		return stream.filter(
-			oAuth2ScopeGrant -> !Collections.disjoint(
-				oAuth2ScopeGrant.getScopeAliasesList(), scopeAliases)
-		).map(
-			oAuth2ScopeGrant -> _scopeLocator.getLiferayOAuth2Scope(
-				oAuth2ScopeGrant.getCompanyId(),
-				oAuth2ScopeGrant.getApplicationName(),
-				oAuth2ScopeGrant.getScope())
-		).filter(
-			Objects::nonNull
-		).filter(
-			liferayOAuth2Scopes::contains
-		).collect(
-			Collectors.toList()
-		);
+		Collection<LiferayOAuth2Scope> finalLiferayOAuth2Scopes =
+			liferayOAuth2Scopes;
+
+		return TransformUtil.transform(
+			_oAuth2ScopeGrantLocalService.getOAuth2ScopeGrants(
+				oAuth2ApplicationScopeAliasesId, QueryUtil.ALL_POS,
+				QueryUtil.ALL_POS, null),
+			oAuth2ScopeGrant -> {
+				if (Collections.disjoint(
+						oAuth2ScopeGrant.getScopeAliasesList(), scopeAliases)) {
+
+					return null;
+				}
+
+				LiferayOAuth2Scope liferayOAuth2Scope =
+					_scopeLocator.getLiferayOAuth2Scope(
+						oAuth2ScopeGrant.getCompanyId(),
+						oAuth2ScopeGrant.getApplicationName(),
+						oAuth2ScopeGrant.getScope());
+
+				if ((liferayOAuth2Scope == null) ||
+					!finalLiferayOAuth2Scopes.contains(liferayOAuth2Scope)) {
```

Comment:
@JanisZhang Please double check this part. If the list is empty, no `liferayOAuth2Scope` will be contained in it, then it will return null for every `oAuth2ScopeGrant`.